bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/FileRecoveryLog.java

```diff
@@ -0,0 +1,237 @@
+     *
+     * @return The latest version of the recovery log file.
+     */
+    private int findLatestVersion() {
+        int latestVersion = 0;
+        File directory = recoveryLogDir.toFile();
+        File[] files = directory.listFiles((dir, name) -> name.matches(baseFileName + "(\\d+)\\.log"));
+        if (files == null) {
+            return latestVersion;
+        }
+        for (File file : files) {
+            String fileName = file.getName();
+            int version = Integer.parseInt(fileName.replaceAll(
+                    baseFileName, "").replaceAll(".log", "")
+            );
+            if (version > latestVersion) {
+                latestVersion = version;
+            }
+        }
+        return latestVersion;
+    }
+
+    /**
+     * Initializes the append channel for the given file.
+     *
+     * @param file The file to initialize the append channel for.
+     */
+    private void initAppendChannel(File file) {
+        try {
+            appendChannel = FileChannel.open(file.toPath(), StandardOpenOption.APPEND);
+            FileLock lock = appendChannel.tryLock();
+            if (lock == null) {
+                stderr.println("error: failed to acquire lock on recovery log file " + file.toPath());
+            }
+        } catch (IOException e) {
+            stderr.println("error: failed to acquire lock on recovery log file " + file.toPath());
+        }
+    }
+
+    @Override
+    public void put(TransactionLogRecord trxRecord) {
+        boolean force = !(trxRecord.getTransactionState().equals(RecoveryState.TERMINATED)); // lazy write for done record
+        writeToFile(trxRecord.getTransactionLogRecord(), force);
+        if (checkpointInterval != -1) {
+            ifNeedWriteCheckpoint();
+            numOfPutsSinceLastCheckpoint++;
+        }
+    }
+
+    public Map<String, TransactionLogRecord> getPendingLogs() {
+        Map<String, TransactionLogRecord> pendingTransactions = new HashMap<>();
+        Map<String, TransactionLogRecord> transactionLogs = readLogsFromFile(file);
+        if (transactionLogs == null) {
+            return null;
+        }
+        for (Map.Entry<String, TransactionLogRecord> entry : transactionLogs.entrySet()) {
+            String trxId = entry.getKey();
+            TransactionLogRecord trxRecord = entry.getValue();
+
+            if (!trxRecord.isCompleted()) {
+                pendingTransactions.put(trxId, trxRecord);
+            }
+        }
+        return pendingTransactions;
+    }
+
+    /**
+     * Write a transaction log entry to the recovery log file.
+     *
+     * @param str the log entry to write
+     */
+    private void writeToFile(String str, boolean force) {
+        if (appendChannel == null || !appendChannel.isOpen()) {
+            initAppendChannel(file);
+        }
+        byte[] bytes = str.getBytes();
+        try {
+            appendChannel.write(java.nio.ByteBuffer.wrap(bytes));
```

Comment:
Can't we import `java.nio.ByteBuffer` and use `ByteBuffer.wrap` here

Reply:
Also see if we can use try-with-resources in cases where you use try-catch with a resource. https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html#:~:text=The%20try%20%2Dwith%2Dresources%20statement%20is%20a%20try%20statement%20that,the%20end%20of%20the%20statement.

Reply:
The intention is to keep the `appendChannel` open for write operations throughout every phase of a 2PC transaction. So, the use of try-with-resources would not be applicable for this scenario.